kcmdline.c

```diff
@@ -0,0 +1,89 @@
+#include "kcmdline.h"
+
+static const CHAR8 allowed[6][12] = {
+	"^console=",
+	"^root=oci:",
+	"root=atomix",
+	"ro",
+	"quiet",
+	"verbose"
+};
+
+#define MAX_TOKENS 128
+
+BOOLEAN is_allowed(const CHAR8 *input) {
+	int len = 0;
+	const CHAR8 * token;
+    int num_allowed = sizeof(allowed) / sizeof(allowed[0]);
+	for (int i=0; i<num_allowed; i++) {
+		token = allowed[i];
+		len = strlena(token);
+		if (token[0] == '^') {
+			len = len - 1;
+			token = &token[1];
+		}
+		if (strncmpa(token, input, len) == 0) {
+			return TRUE;
+		}
+	}
+	return FALSE;
+}
+
+// check cmdline to make sure it contains only allowed words.
+// return EFI_SUCCESS on safe, EFI_SECURITY_VIOLATION on unsafe;
+EFI_STATUS check_cmdline(CONST CHAR8 *cmdline, UINTN cmdline_len) {
+	CHAR8 c = '\0';
+	CHAR8 *buf = NULL;
+	CHAR8 *tokens[MAX_TOKENS];
+	int i;
+	int start = -1;
+	int num_toks = 0;
+	buf = AllocatePool(cmdline_len);
+	if (!buf)
+		return EFI_OUT_OF_RESOURCES;
+
+	CopyMem(buf, cmdline, cmdline_len);
+
+	// walk buf, populating tokens
+	for (i = 0; i < cmdline_len; i++) {
+		if ( start < 0 ) {
+			start = i;
+		}
+
+		c = buf[i];
+		if (c < 0x20 || c > 0x7f) {
+			Print(L"Bad character %x.", buf);
+			return EFI_SECURITY_VIOLATION;
+		}
+		if (i >= MAX_TOKENS) {
+			Print(L"Too many tokens in cmdline.");
+			return EFI_SECURITY_VIOLATION;
+		}
+
+		if (c == ' ' || c == '\0') {
+			if (start >= 0) {
+				tokens[num_toks] = &buf[start];
+				start = -1;
+				num_toks++;
+			}
+			if (c == '\0') {
```

Comment:
The indent here is off, starts with spaces.  (if you look with regular vi. Can't tell in the webview).

Reply:
yeah, needs fixing. i'll fix tomorow